### ISSUE TITLE
SC2317 false positive warning

### ISSUE THREAD

**Issue by orbea** [koalaman/shellcheck#2542]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2317, SC2317
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh

set -e

quit () { err=$?; bar; }

trap 'quit; trap - EXIT; exit $err' EXIT INT

foo

exit 0

```

#### Here's what shellcheck currently says:
```
Line 3	SC2317: Command appears to be unreachable. Check usage (or ignore if invoked indirectly).
Line 3	SC2317: Command appears to be unreachable. Check usage (or ignore if invoked indirectly).
```
#### Here's what I wanted or expected to see:

This should not be a warning, the `exit 0` at the end of a script is very common and the `quit` function is indeed reachable due to the `trap`. In this case the `exit 0` may never be reached if the `foo` command fails for whatever reason and even if the `exit 0` is reached the `quit` function which will have any cleanup is still activated due to the `trap`.

**Comment by dimo414** [comment]
It's the `exit 0` at the end of the script that appears to be triggering this finding. Removing that line clears the "unreachable" finding. Here's a smaller example:

```sh
#!/bin/sh
foo() { bar; }
exit 0  # comment this out to do away with the finding
```

I disagree with your assertion that it's very common" to end a script with `exit 0` - it's basically a no-op, and should rarely be needed, especially when using `set -e`. So a simple fix would be to remove that line from your script. The [wiki](https://www.shellcheck.net/wiki/SC2317) mentions this:

> Defined functions are assumed to be reachable when the script ends (not exits) since another file may source and invoke them.

That said, I agree this finding is confusing and at a minimum it should point to the `exit` line and explain how it's coming to this reachability conclusion.

**Comment by orbea** [comment]
> So a simple fix would be to remove that line from your script. 

You'll need to remove it from thousands of scripts.In other words that is not a reasonable solution.

**Comment by jjakob** [comment]
@dimo414 a trap on EXIT will be invoked even with `exit 0`, so there's no reason to say that function will not be reached.

**Comment by dimo414** [comment]
I agree that the finding is wrong, I'm just saying `exit 0` isn't needed and removing that line appears to make the finding go away.

**Comment by orbea** [comment]
> I'm just saying `exit 0` isn't needed and removing that line appears to make the finding go away.

As already explained this is clearly not the solution.

**Comment by oliv3r** [comment]
> I disagree with your assertion that it's very common" to end a script with `exit 0` - it's basically a no-op, and should rarely be needed, especially when using `set -e`.

I diagree with your assertion too :)

The final `exit 0` at the end, is your canary. E.g. if your script walks its normal life, and exists normally via that exit, it will prevent any attacks such as:

`echo "rm -rf /" >> script.sh`

so its simply a best practice to _always_ do it.

Having said that, you are correct that _technically_ it is not needed.

**Comment by qneill-sifive** [comment]
My script also hits this issue, and it ends with:

```if has_some_error_condition; then
    exit 1
else
    exit 0
fi```

**Comment by wileyhy** [comment]
I am experiencing this issue as well. I've found `exit 0` to be
particularly useful during debugging, as is `exit 00`, `exit 101` and `exit
"$LINENO"`. It is valid syntax.

Thanks,
Wiley

**Comment by ringerc** [comment]
It can't seem to cope with traps like

```
function premature_exit() {
    echo 1>&2 "Unexpected premature exit from entrypoint script at $0:$line"
}

trap 'line=${BASH_LINENO:-};premature_exit' EXIT

# ... do work ...
errcode=$?

# clear trap
trap EXIT

# exit with errcode propagated from inner workload
exit $errcode
```

**Comment by marc-hb** [comment]
Still reproducible with today's shellcheck.net

Even if you don't find `exit 0` useful, you should still find `{ main "$@"; exit $?; }` useful to safely edit the script while it's running some time-consuming test suite (think XKCD 303). Useful trick I just learned at  https://hachyderm.io/@simontatham/114511220670677410

```shell
#!/bin/bash

trap exit_handler EXIT

exit_handler()
{
    printf 'exit handler\n'
}

# Silence SC2317 and SC2329 (depending on the shellcheck version) without any comment.
false && exit_handler "$@"

main()
{
    printf 'main\n'
}

{
  main "$@"; exit $?
}
```

**Comment by marc-hb** [comment]
@orbea could you add "trap" and "EXIT" in the issue title? This should help avoid people filing duplicates like #2660

**Comment by e40** [comment]
> I disagree with your assertion that it's very common" to end a script with `exit 0` - it's basically a no-op, and should rarely be needed, especially when using `set -e`. So a simple fix would be to remove that line from your script. The [wiki](https://www.shellcheck.net/wiki/SC2317) mentions this:

This is absolutely false.  I end ALL shell scripts with it, in this way:

```
{
    stuff...

    exit 0
}
```
and there is a very good reason: if I _don't_ do this, then changing the script while it's running  will result in undefined behavior, often errors.

**Comment by e40** [comment]
I really wish this would be fixed, because I have to disable that for pretty much every script I write (reason for `exit 0` given above).
